element_calcium_imaging/scan.py

```diff
@@ -613,3 +607,125 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
             pathlib.Path(f).relative_to(root_dir).as_posix() for f in scan_filepaths
         ]
         self.ScanFile.insert([{**key, "file_path": f} for f in scan_files])
+
+
+@schema
+class ScanQualityMetrics(dj.Computed):
+    """
+    Metrics to assess quality of Scan.
+    """
+
+    definition = """
+    -> ScanInfo.Field
+    """
+
+    class FrameMetrics(dj.Part):
+        definition = """
+        -> master
+        channel: int
+        ---
+        min_intensity: longblob   # Minimum value of each frame
+        mean_intensity: longblob  # Mean value at each frame
+        max_intensity: longblob   # Maximum value at each frame
+        contrast: longblob        # Contrast of each frame, difference of 99 and 1 percentiles.
+        """
+
+    class QuantalSize(dj.Part):
+        """
+        Quantities inferred from Photon Transfer Curve.
+        """
+
+        definition = """ 
+        -> master
+        channel: int
+        ---
+        min_intensity   : int      # Minimum value in movie.
+        max_intensity   : int      # Maximum value in movie.
+        quantal_size    : float    # Quantal size (variance/mean intensity slope), gain.
+        zero_level      : int      # Level corresponding to zero (computed from variance dependence)
+        quantal_frame   : longblob # Average frame expressed in quanta
+        """
+
+    def make(self, key):
+        acq_software, nchannels = (Scan * ScanInfo & key).fetch1(
+            "acq_software", "nchannels"
+        )
+
+        self.insert1(key)
+
+        for channel in range(nchannels):
+            if acq_software == "ScanImage":
+                import scanreader
+
+                movie = scanreader.read_scan(get_scan_image_files(key))[
+                    key["field_idx"], :, :, channel, :
+                ].transpose(2, 0, 1)
+            elif acq_software == "Scanbox":
+                import sbxreader
+
+                movie = sbxreader(get_scan_box_files(key))[
+                    :, key["field_idx"], channel, :, :
+                ]
+            elif acq_software == "NIS":
+                import nd2
+
+                nd2_file = nd2.ND2File(get_nd2_files(key)[0])
+
+                get_dim_pos = {k: i for i, k in enumerate(nd2_file.sizes)}
+
+                known_fields = ["T", "Z", "C", "Y", "X"]
```

Comment:
a better variable name
```suggestion
                valid_dimensions = ["T", "Z", "C", "Y", "X"]
```

Reply:
Definitely! I like the use of set with strings.

Reply:
@dimitri-yatsenko The use of set here introduced a bug in the code. Apparently set changes the order of a sequence.

Reply:
Yes, sets are unordered.